q2_diversity_lib/_alpha/_method.py

```diff
@@ -32,3 +34,44 @@ def faith_pd(table: biom.Table, phylogeny: skbio.TreeNode) -> pd.Series:
 
     result.name = 'faith_pd'
     return result
+
+
+def observed_features(table: biom.Table) -> pd.Series:
+    if table.is_empty():
```

Comment:
This guard could be be refactored into a [decorator.](https://www.python.org/dev/peps/pep-0318/) While not necessary, it could be worthwhile for your python education.

Reply:
I wish I could heart this more times, @thermokarst. 
Are abstractions like this less desirable in testing for readability reasons?

Reply:
Hmm, good question! I don't think I have any strong thoughts on this one way or the other w.r.t. testing, but, I do think thinking about readability in tests is important! How is that for a non-answer?

Reply:
Word of warning, dealing with the parameters and type annotations (which must be introspected by the framework) will make decorators a bit of a pain to deal with. There is a solution (the same one the framework uses), but it does involve compiling source code on the fly (this is not typically what decorators need to do).

Although, if you are ok with the decorator only being usable by alpha div functions, then I suppose the signature is usually well defined ahead of time and can be baked into the decorator.

Reply:
Actually, the newish `__signature__` protocol might make his more feasible than I am thinking.

Reply:
Luckily we have this in our toolbox https://docs.python.org/3/library/functools.html#functools.wraps

This preserves the signature of the wrapped function!

Reply:
oh right! that should probably work fine, also it looks like there are some new parameters on it, we might be able to ditch the `decorator` package in `action.py` in the framework (obviously super unrelated to this).

Reply:
Wrap works like a champ, @thermokarst. Thanks!